File: src/algorithms/dynamic-programming/1d-linear/count-bits/sources/count-bits.ts
// Count Bits tabulation — dp[i] = number of 1-bits in binary representation of i
function countBits(targetNumber: number): number[] {
  // @step:initialize
  const dpTable = new Array(targetNumber + 1).fill(0); // @step:initialize,fill-table
  // dp[0] = 0: zero has no set bits
  for (let bitIndex = 1; bitIndex <= targetNumber; bitIndex++) {
    // @step:compute-cell
    // Half the number shares all bits except possibly the LSB
    dpTable[bitIndex] = dpTable[bitIndex >> 1] + (bitIndex & 1); // @step:compute-cell,read-cache
  }
  return dpTable; // @step:complete
}

File: src/algorithms/dynamic-programming/count-bits/sources/count-bits.py
# Count Bits tabulation — dp[i] = number of 1-bits in binary representation of i
def count_bits(target_number: int) -> list[int]:  # @step:initialize
    dp_table = [0] * (target_number + 1)  # @step:initialize,fill-table
    # dp[0] = 0: zero has no set bits
    for bit_index in range(1, target_number + 1):  # @step:compute-cell
        # Half the number shares all bits except possibly the LSB
        dp_table[bit_index] = dp_table[bit_index >> 1] + (bit_index & 1)  # @step:compute-cell,read-cache
    return dp_table  # @step:complete
